specification/dsdl/conventions.tex

```diff
@@ -0,0 +1,115 @@
+
+Vendor-specific, user-specific, or any other data types not defined by this specification
+must not be defined inside the standard root namespace.
+Vendor-specific or user-specific root namespaces must be used instead.
+
+\subsection{Naming recommendations}
+
+The DSDL naming recommendations follow those that are widely accepted in the general software development industry.
+
+\begin{itemize}
+    \item Namespaces and field attributes should be named in the \verb|snake_case|.
+    \item Constant attribute should be named in the \verb|SCREAMING_SNAKE_CASE|.
+    \item Data types (excluding their namespaces) should be named in the \verb|PascalCase|.
+    \item Names of message types should form a declarative phrase or a noun. For example,
+    \verb|BatteryStatus| or \verb|OutgoingPacket|.
+    \item Names of service types should form an imperative phrase or a verb. For example,
+    \verb|GetInfo| or \verb|HandleIncomingPacket|.
+    \item Avoid short names, unnecessary abbreviations, and uncommon acronyms.
+\end{itemize}
+
+\subsection{Comments}
+
+Every data type definition file should begin with a header comment that provides an exhaustive description
+of the data type, its purpose, semantics, usage patterns, any related data exchange patterns,
+assumptions, constraints, and all other information that may be necessary or generally useful for the usage of the
+data type definition.
+
+Every attribute of the data type definition, and especially every field attribute of it,
+should have an associated comment explaining the purpose of the attribute, its semantics, usage patterns,
+assumptions, constraints, and any other pertinent information.
+Exception applies to attributes supplied with sufficiently descriptive and unambiguous names.
+
+% Field comment placement https://forum.uavcan.org/t/dsdl-documentation-comments/407
+
+\subsection{Optional value representation}
+
+Data structures may include optional field attributes that are not necessarily always populated.
+
+The recommended approach for representing optional field attributes
+is to use variable-length arrays with the capacity of one element,
+prefixed with padding bits as necessary to retain byte alignment.
+
+Alternatively, such one-element variable-length arrays can be replaced with two-field unions,
+where the first field is empty and the second field contains the desired optional value.
+The described layout is bit-compatible and semantically compatible with the one-element array described above,
+provided that the field attributes are not swapped.
+
+Floating-point-typed field attributes may be assigned the value of not-a-number (NaN) per IEEE 754
+to indicate that the value is not specified;
+however, this pattern is discouraged because some nodes may not support IEEE 754 NaN values internally,
+the value would still have to be transferred over the bus even if not populated,
+and special case values undermine type safety.
```

Comment:
I've seen real defects in the field caused by mixed NaN support. Should we take a harder stance in the spec and require all nodes support it or that it is not supported/allowed?

Reply:
I'm worried that might be an overreach for this specification. Instead, I suggest we add a blue box right under this last sentence briefly explaining the dangers of mixed NaN interpretation. Do you have any particular wording in mind?

Reply:
(Channeling a colleague of mine) shouldn't we just support IEEE 754 with no caveats? We wouldn't be the ones reaching. IEEE 754 is the standard that is reaching but its pretty well established (what, like 1985, right?).

(switching back to me) What known microprocessors and/or softfp implementations don't support NaN but do support IEEE 754?

Reply:
> shouldn't we just support IEEE 754 with no caveats?

But that's exactly what we do. I don't understand.

> What known microprocessors and/or softfp implementations don't support NaN but do support IEEE 754?

I don't know. Should we care? We refer to IEEE754 mostly for the sake of its well-defined bit-level representations. Platforms may internally use whatever floating point formats they want, it's none of our business, provided that they transform their internal formats into/from IEEE754 when serializing/deserializing.

Reply:
> But that's exactly what we do. I don't understand.

The caveat is "this pattern is discouraged because some nodes may not support IEEE 754 NaN". As far as I know NaN support isn't optional for IEEE754, correct?

Reply:
But support for IEEE 754 *is* very optional; hence support for NaN is also optional.

Reply:
> IEEE 754 is very optional

Oh, sorry. Of course. So backing up then: mixed NaN support is a s**t-show. We should define and require support for a NaN value so nodes that do support IEEE754 know how to transmit them on the bus. Nodes that do not support IEEE754 will have to have a special case to handle them which is better then undefined behaviour. Perhaps we should add `sNaN` and `qNaN` as special, rational values in the language?

Reply:
How about this: I will add a few sentences requiring that platforms that do not natively support IEEE 754 must be able to accommodate IEEE 754 NaN values. This should probably go into "3.4.2 Primitive types" with an additional reference in "3.7.3.4 Floating point types". I don't think we should add NaN support for constant expressions yet, but this may be changed later without compromising backward compatibility.

Reply:
but if I have a NaN value and I want to send it on the bus what do I do?

Reply:
You just send it. I am saying that we do not seem to need NaN values in DSDL.

```python
float16 a  # you can assign NaN to this at runtime
float16 A = NaN  # Can't do this
```

Reply:
What about in dsdl expressions. What happens if we end up with something that is normally NaN like division by zero?

Reply:
We get a DSDL-processing-time error. The spec says that the division (and modulo) operator is not defined for x/0, so the DSDL front-end would just report an error.

Reply:
Fixed in https://github.com/UAVCAN/specification/pull/46/commits/f9ec8842207647366d35bc8d02a8f3c702f1da5d